### PR TITLE
Update to wgpu-rs 0.7

### PR DESCRIPTION
**Pull request by Neo-Zhixing**
This PR makes bevy's renderer compatible with upstream wgpu-rs .

Wait to merge this until wgpu-rs's next release.

### CODE REVIEW

File: crates/bevy_wgpu/src/renderer/wgpu_render_resource_context.rs [hunk 1]
```diff
@@ -480,7 +482,23 @@ impl RenderResourceContext for WgpuRenderResourceContext {
...
+                                Bound::Included(&bound) => BufferSize::new(bound + 1 - offset),
+                                Bound::Excluded(&bound) => BufferSize::new(bound - offset),
+                                Bound::Unbounded => None,
+                            };
+
+                            wgpu::BindingResource::Buffer {
+                                buffer: wgpu_buffer,
+                                offset,
+                                size,
+                            }
```
Comment (Neo-Zhixing) [inline crates/bevy_wgpu/src/renderer/wgpu_render_resource_context.rs:503]: Not sure if this is a good way to do this. With gfx-rs/wgpu-rs#523 we've got a new method .as_entire_binding(), but it takes the entire buffer as a biding resource as opposed to a range of the buffer.

This code was copied from `wgpu::range_to_offset_size` but it is private.

File: crates/bevy_render/src/pipeline/pipeline_compiler.rs [hunk 2]
```diff
@@ -15,16 +15,16 @@ pub struct PipelineSpecialization {
     pub shader_specialization: ShaderSpecialization,
     pub primitive_topology: PrimitiveTopology,
     pub dynamic_bindings: HashSet<String>,
-    pub index_format: IndexFormat,
-    pub vertex_buffer_descriptor: VertexBufferDescriptor,
+    pub index_format: Option<IndexFormat>,
+    pub vertex_buffer_descriptor: VertexBufferLayout,
```
Comment (cart) [inline crates/bevy_render/src/pipeline/pipeline_compiler.rs:19]: nit: lets rename this to vertex_buffer_layout

File: crates/bevy_render/src/pipeline/pipeline_layout.rs [hunk 3]
```diff
@@ -49,7 +49,7 @@ impl PipelineLayout {
             }
         }
 
-        for vertex_buffer_descriptor in shader_layouts[0].vertex_buffer_descriptors.iter() {
+        for vertex_buffer_descriptor in shader_layouts[0].vertex_buffer_layouts.iter() {
```
Comment (cart) [inline crates/bevy_render/src/pipeline/pipeline_layout.rs:52]: nit: vertex_buffer_layout

File: crates/bevy_wgpu/src/renderer/wgpu_render_resource_context.rs (hunk 1 above)
Comment (cart) [reply crates/bevy_wgpu/src/renderer/wgpu_render_resource_context.rs:565]: the current code looks good to me!

**Review by cart — COMMENTED** [review]
Looks good to me. Thanks for doing this. I pushed one small change removing PipelineDescriptor::index_format. strip_index_format has replaced it. We now specialize on that.

File: crates/bevy_render/src/shader/shader_reflect.rs [hunk 4]
```diff
@@ -123,27 +122,34 @@ fn reflect_binding(
...
             },
         ),
         ReflectDescriptorType::SampledImage => (
             &binding.name,
-            BindType::SampledTexture {
-                dimension: reflect_dimension(type_description),
-                component_type: TextureComponentType::Float,
+            BindType::Texture {
+                view_dimension: reflect_dimension(type_description),
+                sample_type: TextureSampleType::Float { filterable: false },
```
[the remark this answers is not in the document]
Comment (Neo-Zhixing) [reply crates/bevy_render/src/shader/shader_reflect.rs]: Yeah I guess default to filterable textures make sense.

File: crates/bevy_wgpu/src/wgpu_renderer.rs [hunk 5]
```diff
@@ -37,7 +37,7 @@ impl WgpuRenderer {
             .request_adapter(&wgpu::RequestAdapterOptions {
                 power_preference: match options.power_pref {
                     WgpuPowerOptions::HighPerformance => wgpu::PowerPreference::HighPerformance,
-                    WgpuPowerOptions::Adaptive => wgpu::PowerPreference::Default,
+                    WgpuPowerOptions::Adaptive => Default::default(),
```
[the remark this answers is not in the document]
Comment (Neo-Zhixing) [reply crates/bevy_wgpu/src/wgpu_renderer.rs]: Now defaults to LowPower.

**Comment by Neo-Zhixing** [comment]
wgpu 0.7 was just released! 🎉🎉🎉🎉

This PR is now ready to be merged.

**Comment by cart** [comment]
Awesome!

**Pull request merged**